p2p/protocol/identify/obsaddr.go

```diff
@@ -213,6 +213,36 @@ func (o *ObservedAddrManager) AddrsFor(addr ma.Multiaddr) (addrs []ma.Multiaddr)
 	return res
 }
 
+// appendInferredAddrs infers the external address of other transports that
+// share the local thin waist with a transport that we have do observations for.
+//
+// e.g. If we have observations for a QUIC address on port 9000, and we are
+// listening on the same interface and port 9000 for WebTransport, we can infer
+// the external WebTransport address.
+func (o *ObservedAddrManager) appendInferredAddrs(twToObserverSets map[string][]*observerSet, addrs []ma.Multiaddr) []ma.Multiaddr {
+	if twToObserverSets == nil {
+		twToObserverSets = make(map[string][]*observerSet)
+		for localTWStr := range o.externalAddrs {
+			twToObserverSets[localTWStr] = append(twToObserverSets[localTWStr], o.getTopExternalAddrs(localTWStr)...)
+		}
+	}
+	for _, a := range o.listenAddrs() {
```

Comment:
Yeah. I'm now doing it on both listenAddrs and interfaceListenAddrs. I think in the ideal world we would just use `interfaceListenAddrs` because those are what map to the 4-tuple for connection flows. For UDP right now, this doesn't work yet as you say. So we have to do it for both because the 0.0.0.0 is what shows up in the observed address mapping due to it being the conn.LocalMultiAddr result.